src/PubSubClient.cpp

```diff
@@ -453,46 +453,15 @@ bool PubSubClient::publish_P(const char* topic, const char* payload, bool retain
 }
 
 bool PubSubClient::publish_P(const char* topic, const uint8_t* payload, size_t plength, bool retained) {
-    uint8_t llen = 0;
-    uint8_t digit;
-    size_t rc = 0;
-    size_t tlen;
-    size_t pos = 0;
-    const uint8_t header = MQTTPUBLISH | (retained ? MQTTRETAINED : 0);
-    size_t len;
-    size_t expectedLength;
-
-    if (!connected()) {
-        return false;
-    }
-
-    tlen = strnlen(topic, this->bufferSize);
-
-    this->buffer[pos++] = header;
-    len = plength + 2 + tlen;
-    do {
-        digit = len & 127;  // digit = len % 128
-        len >>= 7;          // len = len / 128
-        if (len > 0) {
-            digit |= 0x80;
+    if (beginPublish(topic, plength, retained)) {
+        size_t rc = 0;
+        for (size_t i = 0; i < plength; i++) {
+            rc += _client->write((uint8_t)pgm_read_byte_near(payload + i));
```

Comment:
This is rather slow, at least on ESP8266 as this may read every pgm byte 4 times.
Maybe for some other PR to look into making this faster, like how it is done in the esp8266/Arduino `String` class.
N.B. for ESP32 it will be different, so that's why it is something for a new PR to look into.

Reply:
Yes, I already tried
`size_t rc = _client->write_P((PGM_P)payload, plength);`
But this won't work using Ethernet client and boards != ESP8266/ESP32 ...

Reply:
You could add a platform specific define at the top of the .cpp file.
```c++
#if defined(ESP8266) || defined(ESP32)
  #define CLIENT_WRITE_P  _client->write_P
#else
  #define CLIENT_WRITE_P  _client->write
#endif
```

Something like that.

Reply:
I'm sorry. I was wrong. `write_P()` is not declared in class `Client`. It's in my used clients `WiFiClient` and `WiFiClientSecure`. But the base is `Client`/ `Print`. So using `_client->write_P()` won't work at all.